source/material_model/steinberger.cc

```diff
@@ -383,6 +427,77 @@ namespace aspect
     }
 
 
+    template <int dim>
+    std::vector<std::vector<double>>
+                                  Steinberger<dim>::
+                                  phase_volume_fractions (const MaterialModel::MaterialModelInputs<dim> &in) const
+    {
+      // In the following function,
+      // the index i corresponds to the ith compositional field
+      // the index j corresponds to the jth unique phase
+      // the index k corresponds to the kth evaluation point
+
+      std::vector<std::vector<double>> volume_fractions(unique_phase_names.size(),
+                                                        std::vector<double>(in.n_evaluation_points(), 0.));
+
+      if (material_lookup.size() == 1)
+        {
+          // if there is only one lookup, unique_phase_names is in the same order as the lookup
+          for (unsigned int j = 0; j < unique_phase_names.size(); j++)
+            {
+              for (unsigned int k = 0; k < in.n_evaluation_points(); k++)
+                volume_fractions[j][k] += material_lookup[0]->phase_volume_fraction(j,in.temperature[k],in.pressure[k]);
+            }
+        }
+      else if (material_lookup.size() == in.composition[0].size() + 1) // background field
+        {
+          // The first material lookup corresponds to a background composition
+          // We first look up the volume fractions corresponding to this background field
+          // (assuming the domain is filled 100% with this single composition)
+          std::vector<std::vector<double>> background_volume_fractions(unique_phase_names.size(),
+                                                                       std::vector<double>(in.n_evaluation_points(), 0.));
+          for (unsigned int j = 0; j < unique_phase_indices[0].size(); j++)
+            {
+              for (unsigned int k = 0; k < in.n_evaluation_points(); k++)
+                {
+                  background_volume_fractions[unique_phase_indices[0][j]][k] += material_lookup[0]->phase_volume_fraction(j,in.temperature[k],in.pressure[k]);
+                  volume_fractions[unique_phase_indices[0][j]][k] += background_volume_fractions[unique_phase_indices[0][j]][k];
```

Comment:
I think you can avoid looping over j and k twice, if you loop from i=0 below and then just make this the spacial case for i=0.

Reply:
Ok!